c2rust-analyze/src/main.rs

```diff
@@ -465,7 +514,22 @@ fn run(tcx: TyCtxt) {
                         }
                         _ => continue,
                     },
-                    // TODO: Rvalue::Cast (certain cases)
+                    Rvalue::Cast(_, ref op, _) => {
+                        if let Some(p) = op.place().filter(|p| acx.c_void_ptrs.contains(p)) {
+                            // This is a special case for types being casted from *c_void to a pointer
+                            // to some other type, e.g. `let foo = malloc(..) as *mut Foo;`
+                            // carry over the pointer id of *c_void, but match the pointer ids
+                            // of the casted-to-type for the rest
+                            acx.c_void_casts.0.insert(p, lhs);
+                        }
+
+                        if acx.c_void_ptrs.contains(&lhs) {
+                            // This is a special case for types being casted to *c_void
+                            acx.c_void_casts.0.insert(lhs, op.place().unwrap());
+                        }
+
+                        continue;
+                    }
```

Comment:
The `Operand` of an `Rvalue::Cast` is never `Operand::Constant`, right?  Then we can just

```rust
let p = op.place().expect("Rvalue::Cast is not Operant::Constant so it has a Place");
```

which simplifies things and makes the code's intent more clear (right now there's one `if let Some` and one `.unwrap()`).

Reply:
Can you extract out the

```rust
if self.ptrs.contains(&p1) {
    self.casts.0.insert(p1, p2);
}
```

part of this into some `struct CVoids`?  Then it's clear the logic is the same for both, just flipped.

Reply:
`expect` added in 1d95fed5. will delay extracting the other code into a method -- that pattern only shows up twice AFAICT.